### DISCUSSION TITLE
Put Sparse Matrix on CUDA

### DISCUSSION THREAD

**Discussion by RX28666** [pyg-team/pytorch_geometric#5966] [Q&A]
Hi, I am trying to compute the normalized adjacency matrix `D(-1/2)(A+I)D(-1/2)` on Cuda, this is what I am doing:
```
adj = torch_geometric.utils.to_scipy_sparse_matrix(data.edge_index, num_nodes = data.num_nodes)
M_test = calc_A_hat(adj_test)
A_test = sparse_matrix_to_torch((1 - args.alpha) * M_test)
A_test = A_test.to(device)
```
Where:
```
def calc_A_hat(adj_matrix: sp.spmatrix) -> sp.spmatrix:
       nnodes = adj_matrix.shape[0]
       A = adj_matrix + sp.eye(nnodes)
       D_vec = np.sum(A, axis=1).A1
       D_vec_invsqrt_corr = 1 / np.sqrt(D_vec)
       D_invsqrt_corr = sp.diags(D_vec_invsqrt_corr)
       return D_invsqrt_corr @ A @ D_invsqrt_corr
```
and:
```
def sparse_matrix_to_torch(X):
    coo = X.tocoo()
    indices = np.array([coo.row, coo.col])
    return torch.sparse.FloatTensor(torch.LongTensor(indices), torch.FloatTensor(coo.data), coo.shape)
```
It seems all calculations are done on CPU, I tried to put `adj` to Cuda first by `adj = adj.to(device)`, but it returned `to not found` in sparse matrix, I am wondering how I can put the sparse matrix on CUDA so I can do this calculations on GPU, or there is another way to get the normalized adjacency matrix on CUDA, because the calculations are very slow when the graph is large. Any help would be appreciated.

**Comment by EdisonLeeeee** [comment]
You were using `scipy` sparse matrix that is not supported on CUDA. It's suggested to use `torch_sparse.SparseTensor` instead, see https://github.com/pyg-team/pytorch_geometric/blob/32c6b7abb7ba4c1c964350af7d70f4ca9245fe7a/torch_geometric/nn/conv/gcn_conv.py#L32-L34
